File: Sorting/p15.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class p15 {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        if (nums == null || nums.length < 3)
            return ans;
        Arrays.sort(nums);
        int n = nums.length;
        for (int first = 0; first < n - 2; ++first) {
            if (nums[first] > 0)
                break;
            if (first > 0 && nums[first] == nums[first - 1])
                continue;
            int second = first + 1, third = n - 1;
            int target = -nums[first];
            while (second < third) {
                int sum = nums[second] + nums[third];
                if (sum == target){
                    ans.add(new ArrayList<>(Arrays.asList(nums[first], nums[second++], nums[third--])));
                    while(second<third&&nums[second]==nums[second-1])
                        ++second;
                    while(second<third&&nums[third]==nums[third+1])
                        --third;
                }
                else if (sum > target)
                    --third;
                else
                    ++second;
            }
        }
        return ans;
    }
}

File: Hot-100/Python-Quickpass/p15.py
from typing import List


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        if n < 3:
            return []
        nums = sorted(nums)
        ans = []
        for first in range(n-2):
            if nums[first] > 0:
                break
            if first > 0 and nums[first] == nums[first-1]:
                continue
            second, third = first+1, n-1
            maxSum = nums[third]+nums[third-1]
            if maxSum < -nums[first]:
                continue
            minSum = nums[second]+nums[second+1]
            if minSum > -nums[first]:
                break
            while second < third:
                sum = nums[second]+nums[third]
                if sum == -nums[first]:
                    ans.append([nums[first], nums[second], nums[third]])
                    second += 1
                    third -= 1
                    while second < third and nums[second] == nums[second-1]:
                        second += 1
                    while second < third and nums[third] == nums[third+1]:
                        third -= 1
                elif sum < -nums[first]:
                    second += 1
                else:
                    third -= 1
        return ans


Solution().threeSum([-1, 0, 1])
